pkg/internal/util/patch/resource.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package patch
+
+import (
+	"io"
+	"io/ioutil"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch"
+
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/yaml"
+
+	"sigs.k8s.io/kind/pkg/errors"
+)
+
+type resource struct {
+	raw       string    // the original raw data
+	json      []byte    // the processed data (in JSON form), may be mutated
+	matchInfo matchInfo // for matching patches
+}
+
+func (r *resource) apply6902Patch(patch json6902Patch) (matches bool, err error) {
+	if !r.matches(patch.matchInfo) {
+		return false, nil
+	}
+	patched, err := patch.patch.Apply(r.json)
+	if err != nil {
+		return true, errors.WithStack(err)
+	}
+	r.json = patched
+	return true, nil
+}
+
+func (r *resource) applyMergePatch(patch mergePatch) (matches bool, err error) {
+	if !r.matches(patch.matchInfo) {
+		return false, nil
+	}
+	patched, err := jsonpatch.MergePatch(r.json, patch.json)
+	if err != nil {
+		return true, errors.WithStack(err)
+	}
+	r.json = patched
+	return true, nil
+}
+
+func (r resource) matches(o matchInfo) bool {
+	m := &r.matchInfo
+	// we require kind to match, but if the patch does not specify
+	// APIVersion we ignore it (eg to allow trivial patches across kubeadm versions)
```

Comment:
what are the tradeoffs of this decision?

Reply:
It makes it easier to write kind configs with patches.
I don't see any obvious downside to our use case, it's up to the user to be more specific if they want.

The bigger change is that we don't require all patches to match something, but again, that's up to the user, and actually enables writing a config that patches all versions...

This is more relaxed behavior, and easier to work with.

Reply:
Kubernetes types are identified by (`kind`, `apiVersion`), the `apiVersion` is just that, the API Version. Many versions are near identical. Different kinds are not expected to be.

For our specific use case, I'm not aware of any useful patch you could write that would apply to all objects. However, it would be useful to patch identical fields across kubeadm API versions.